packages/grid/_modules_/grid/GridOverlays.tsx

```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
+import { GridApiContext } from './components/GridApiContext';
+import { GridRootPropsContext } from './context/GridRootPropsContext';
+import { useGridSelector } from './hooks/features/core/useGridSelector';
+import { visibleGridRowCountSelector } from './hooks/features/filter/gridFilterSelector';
+import { gridRowCountSelector } from './hooks/features/rows/gridRowsSelector';
+
+export function GridOverlays() {
+  const apiRef = React.useContext(GridApiContext)!;
+  const props = React.useContext(GridRootPropsContext)!;
+
+  const totalRowCount = useGridSelector(apiRef, gridRowCountSelector);
+  const visibleRowCount = useGridSelector(apiRef, visibleGridRowCountSelector);
+
+  const showNoRowsOverlay = !props.loading && totalRowCount === 0;
+  const showNoResultsOverlay = !props.loading && totalRowCount > 0 && visibleRowCount === 0;
+
+  return (
+    <React.Fragment>
```

Comment:
Can we remove the fragment?

Reply:
How would it work?

Reply:
```diff
diff --git a/packages/grid/_modules_/grid/GridOverlays.tsx b/packages/grid/_modules_/grid/GridOverlays.tsx
index 084a5443..b1fb8c22 100644
--- a/packages/grid/_modules_/grid/GridOverlays.tsx
+++ b/packages/grid/_modules_/grid/GridOverlays.tsx
@@ -12,20 +12,19 @@ export function GridOverlays() {
   const totalRowCount = useGridSelector(apiRef, gridRowCountSelector);
   const visibleRowCount = useGridSelector(apiRef, visibleGridRowCountSelector);

-  const showNoRowsOverlay = !props.loading && totalRowCount === 0;
-  const showNoResultsOverlay = !props.loading && totalRowCount > 0 && visibleRowCount === 0;
+  if (props.loading) {
+    return <apiRef.current.components.LoadingOverlay {...props.componentsProps?.loadingOverlay} />;
+  }

-  return (
-    <React.Fragment>
-      {showNoRowsOverlay && (
-        <apiRef.current.components.NoRowsOverlay {...props.componentsProps?.noRowsOverlay} />
-      )}
-      {showNoResultsOverlay && (
-        <apiRef.current.components.NoResultsOverlay {...props.componentsProps?.noResultsOverlay} />
-      )}
-      {props.loading && (
-        <apiRef.current.components.LoadingOverlay {...props.componentsProps?.loadingOverlay} />
-      )}
-    </React.Fragment>
-  );
+  if (totalRowCount === 0) {
+    return <apiRef.current.components.NoRowsOverlay {...props.componentsProps?.noRowsOverlay} />;
+  }
+
+  if (visibleRowCount === 0) {
+    return (
+      <apiRef.current.components.NoResultsOverlay {...props.componentsProps?.noResultsOverlay} />
+    );
+  }
+
+  return null;
 }
```

Reply:
```diff
diff --git a/packages/grid/_modules_/grid/GridOverlays.tsx b/packages/grid/_modules_/grid/GridOverlays.tsx
index 084a5443..b1fb8c22 100644
--- a/packages/grid/_modules_/grid/GridOverlays.tsx
+++ b/packages/grid/_modules_/grid/GridOverlays.tsx
@@ -12,20 +12,19 @@ export function GridOverlays() {
   const totalRowCount = useGridSelector(apiRef, gridRowCountSelector);
   const visibleRowCount = useGridSelector(apiRef, visibleGridRowCountSelector);

-  const showNoRowsOverlay = !props.loading && totalRowCount === 0;
-  const showNoResultsOverlay = !props.loading && totalRowCount > 0 && visibleRowCount === 0;
+  if (props.loading) {
+    return <apiRef.current.components.LoadingOverlay {...props.componentsProps?.loadingOverlay} />;
+  }

-  return (
-    <React.Fragment>
-      {showNoRowsOverlay && (
-        <apiRef.current.components.NoRowsOverlay {...props.componentsProps?.noRowsOverlay} />
-      )}
-      {showNoResultsOverlay && (
-        <apiRef.current.components.NoResultsOverlay {...props.componentsProps?.noResultsOverlay} />
-      )}
-      {props.loading && (
-        <apiRef.current.components.LoadingOverlay {...props.componentsProps?.loadingOverlay} />
-      )}
-    </React.Fragment>
-  );
+  if (totalRowCount === 0) {
+    return <apiRef.current.components.NoRowsOverlay {...props.componentsProps?.noRowsOverlay} />;
+  }
+
+  if (visibleRowCount === 0) {
+    return (
+      <apiRef.current.components.NoResultsOverlay {...props.componentsProps?.noResultsOverlay} />
+    );
+  }
+
+  return null;
 }
```

Reply:
Oh yeah, I love that, it prevents rendering two different overlay by mistake.

Reply:
Hummm, wait, don't we want loading to display on top of no rows?

Reply:
Currently, neither the no rows overlay nor the no results overlay is displayed if `loading` is true, so my suggestion is equivalent. I don't think it would be cool to display "No rows" and the loader at the same time, both are aligned to the center.

Reply:
> both are aligned to the center

Unless we rework the default UI, e.g. https://material-ui.com/components/data-grid/components/#loading-overlay